groth16/src/api.rs

```diff
@@ -88,6 +124,40 @@ pub fn groth16_setup(
     Ok(())
 }
 
+#[cfg(any(feature = "cuda", feature = "opencl"))]
+#[allow(clippy::large_enum_variant)]
+pub enum SetupResult {
+    BLS12381(
+        CircomCircuit<Scalar>,
+        Parameters<Bls12>,
+        VerifyingKey<Bls12>,
+    ),
+}
+
+#[cfg(any(feature = "cuda", feature = "opencl"))]
+pub fn groth16_setup_and_cache(
+    curve_type: &str,
+    circuit_file: &str,
+    pk_file: &str,
+    vk_file: &str,
+    to_hex: bool,
+) -> Result<SetupResult> {
+    let mut rng = rand::thread_rng();
+    let result = match curve_type {
+        "BLS12381" => {
+            let circuit = create_circuit_from_file::<Scalar>(circuit_file, None);
+            let (pk, vk): (Parameters<Bls12>, VerifyingKey<Bls12>) =
+                Groth16::circuit_specific_setup(circuit.clone(), &mut rng)?;
+            write_pk_vk_to_files(curve_type, pk.clone(), vk.clone(), pk_file, vk_file, to_hex)?;
```

Comment:
@ibmp33 if we have return the whole structure, we don't have to write them into local file.  Meanwhile, you can change the `groth16_setup_and_cache` to `groth16_setup_inplace` and remove the arguments xx_file.